mrmShared/linux/uio_mrf.c

```diff
@@ -814,17 +820,85 @@ mrf_remove(struct pci_dev *dev)
         pci_release_regions(dev);
         pci_disable_device(dev);
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0)
         kzfree(priv);
+#else
+        kfree_sensitive(priv);
+#endif
 
         dev_info(&dev->dev, "MRF Cleaned up\n");
 }
 
 
+/**
+ * AER handling functions.
+ * see https://www.kernel.org/doc/html/latest/PCI/pcieaer-howto.html#non-correctable-non-fatal-and-fatal-errors
+ */
+static
+pci_ers_result_t
+mrf_error_detected(struct pci_dev *dev, pci_channel_state_t state)
+{
+    if (state == pci_channel_io_normal) {
+        /* FIXME: Anything else to do here? */
+        return PCI_ERS_RESULT_CAN_RECOVER;
+    } else if (state == pci_channel_io_frozen) {
+        pci_disable_device(dev);
+        /* FIXME: Anything else to do here? */
+        return PCI_ERS_RESULT_NEED_RESET;
+    } else if (state == pci_channel_io_perm_failure) {
+        return PCI_ERS_RESULT_DISCONNECT;
+    }
+
+    return PCI_ERS_RESULT_NONE;
+}
+
+static
+pci_ers_result_t
+mrf_slot_reset(struct pci_dev *dev)
+{
+	if (pci_enable_device(dev)) {
+		dev_err(&dev->dev, "failed to re-enable after slot reset\n");
+		return PCI_ERS_RESULT_DISCONNECT;
+	}
+
+    pci_restore_state(dev);
```

Comment:
Is a call to `pci_set_master()` needed here?

Reply:
Now I'm less sure.  Looking at the source of `pci_restore_state()`, it seems to be restoring `PCI_COMMAND`.

Reply:
Ah, but `pci_set_master()` **may** be doing more than just setting then just setting `PCI_COMMAND_MASTER`.  I think I've gone far enough down this rabbit hole for today.

Reply:
Let me check that with the real HW I have here (after the recovery is finished). According to the https://github.com/epics-modules/mrfioc2/blob/1a0112b3caa05f7e3e4bfc7c3a30ef356ff3cfc9/mrmShared/linux/uio_mrf.c#L598 we want this to be set.

Reply:
I see that the EVR remains a bus master even after the recovery:
```
[dev@bd-cpu18 ~]$ sudo lspci -s 09:00.0 -vv
09:00.0 Signal processing controller: Xilinx Corporation Device 7011
	Subsystem: Device 1a3e:132c
	Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
```